egs2/TEMPLATE/asr1/pyscripts/feats/ssl_feature_utils.py

```diff
@@ -39,30 +41,61 @@ def format_feature_conf_str(feature_conf: str):
     return feature_conf
 
 
+def build_data_iterator(
+    rspecifier: str,
+    in_filetype: str,
+    utt2num_samples: str,
+    batch_bins: Optional[int] = 1,
+):
+    dataset = ESPnetDataset(
+        [(rspecifier[4:], "speech", in_filetype)],
+        preprocess=None,
+    )
+    sampler = NumElementsBatchSampler(
+        batch_bins=batch_bins,
+        shape_files=[utt2num_samples],
+    )
+    batches = list(sampler)
+    iterator = SequenceIterFactory(
+        dataset=dataset,
+        batches=batches,
+        collate_fn=CommonCollateFn(float_pad_value=0.0, int_pad_value=-1),
+        num_workers=2,
+    ).build_iter(0)
```

Comment:
Is there a specific reason of using numel sampler and sequence iterator, I thought streaming iterator would be enough (it is also easier to trace as it does not go through the sampler)

Reply:
Not really. I just think it is too complicated to define multiple choices of iterators / samplers. So I randomly chose one. Not sure which one is good enough. But in the end, we may find one that covers most of the use cases, which may be enough, right?

Reply:
The reason for using numel sampler and sequence iterator instead of streaming iterator is because the utterances are sorted, thus less computations are wasted due to padding. Another reason for using numel sampler is that it is easier to avoid OOM error than others, e.g. sorted sampler.

Reply:
That's fair, thanks for the clarification.